chore!: adopt log/slog, remove go-kit/log

tjhop opened the pull request:
For: #14355

This commit updates Prometheus to adopt stdlib's log/slog package in
favor of go-kit/log. As part of converting to use slog, several other
related changes are required to get prometheus working, including:
- removed unused logging util func `RateLimit()`
- forward ported the util/logging/Deduper logging by implementing a small custom slog.Handler that does the deduping before chaining log calls to the underlying real slog.Logger
- move some of the json file logging functionality to use prom/common package functionality
- refactored some of the new json file logging for scraping
- changes to promql.QueryLogger interface to swap out logging methods for relevant slog sugar wrappers
- updated lots of tests that used/replicated custom logging functionality, attempting to keep the logical goal of the tests consistent after the transition
- added a healthy amount of `if logger == nil { $makeLogger }` type conditional checks amongst various functions where none were provided -- old code that used the go-kit/log.Logger interface had several places where there were nil references when trying to use functions like `With()` to add keyvals on the new *slog.Logger type

tjhop commented:
@SuperQ if you don't mind taking a look when you have time as well, I'd appreciate it!

And sorry all for the massive change set :sweat_smile:

tjhop commented:
Opening as draft for now so it's known that there's work in progress -- still need to fix up tests around the custom query log testing that's done in `cmd/prometheus/query_log_test.go`, but I believe that should cover most things. 

It builds and has been running well for me in my testing so far; feedback welcome!

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -170,85 +169,91 @@ type flagConfig struct {
 					c.autoReloadInterval, _ = model.ParseDuration("1s")
 				}
-				level.Info(logger).Log("msg", fmt.Sprintf("Enabled automatic configuration file reloading. Checking for configuration changes every %s.", c.autoReloadInterval))
+				logger.Info("Enabled automatic configuration file reloading. Checking for configuration changes every reload interval.", "interval", c.autoReloadInterval)
```
```suggestion
				logger.Info("Enabled automatic configuration file reloading. Checking for configuration changes every", "interval", c.autoReloadInterval)
```

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -170,85 +169,91 @@ type flagConfig struct {
 				c.promqlEnableDelayedNameRemoval = true
-				level.Info(logger).Log("msg", "Experimental PromQL delayed name removal enabled.")
+				logger.Info("Experimental PromQL delayed name removal enabled.")
+			case "utf8-names":
+				model.NameValidationScheme = model.UTF8Validation
+				logger.Info("Experimental UTF-8 support enabled")
```
some merge issue?

tjhop replied:
Yup, that's my bad -- this change set was originally branched from `release-3.0` a ways back and after the merge to main I had to rebase forward some ~150 commits. This accidentally made it's way back in. I'll clean it up. 👍

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -544,20 +550,20 @@ func main() {
 	// Throw error for invalid config before starting other components.
 	var cfgFile *config.Config
-	if cfgFile, err = config.LoadFile(cfg.configFile, agentMode, false, log.NewNopLogger()); err != nil {
+	if cfgFile, err = config.LoadFile(cfg.configFile, agentMode, false, slog.New(slog.NewTextHandler(io.Discard, nil))); err != nil {
```
Suggest creating a utility function for `slog.New(slog.NewTextHandler(io.Discard, nil))` which occurs 147 times in this PR.

tjhop replied:
Sure, that's reasonable. 

Would you rather that live here in Prometheus proper or perhaps as a utility function in prometheus/common (`promslog.NewNopLogger`)?

bboreham replied:
No strong opinions. If there is a package that a large part of the 147 cases are already importing I’d look at using that, but don’t make it weird.

tjhop replied:
https://github.com/prometheus/common/pull/697

feedback welcome!

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -621,11 +627,7 @@ func main() {
 	noStepSubqueryInterval := &safePromQLNoStepSubqueryInterval{}
 	noStepSubqueryInterval.Set(config.DefaultGlobalConfig.EvaluationInterval)
 
-	// Above level 6, the k8s client would log bearer tokens in clear-text.
-	klog.ClampLevel(6)
```
Seem to have lost this line.

tjhop replied:
So `ClampLevel()` isn't a native klog function, but rather a function from the library prometheus replaces klog with:
https://github.com/prometheus/prometheus/blob/main/go.mod#L210-L213
https://pkg.go.dev/github.com/simonpasquier/klog-gokit#ClampLevel
https://pkg.go.dev/github.com/simonpasquier/klog-gokit/v3#ClampLevel

A similar change was made to the prometheus operator when they switched to slog:
https://github.com/prometheus-operator/prometheus-operator/pull/6902/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L147
https://github.com/prometheus-operator/prometheus-operator/pull/6902/files#diff-3a86c514d2d3f52dc7235e0444e3a2c54960795d7e73cecb131b876ce306442bL67-L69

To the best of my current understanding, this is intentional

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -621,11 +627,7 @@ func main() {
-	// Above level 6, the k8s client would log bearer tokens in clear-text.
-	klog.ClampLevel(6)
-	klog.SetLogger(log.With(logger, "component", "k8s_client_runtime"))
-	klogv2.ClampLevel(6)
```
And this one.

tjhop replied:
Same response as the other one -- though I do need to double check and see if/where we still need klog v1 in prometheus to see about creating SetSlogLogger() type redirect func like there is in klogv2 upstream

tjhop replied:
yea, we still have v1 things in discovery, I took inspiration from klog's upstream v1/v2 coexistence docs and implemented a small wrapper to redirect klogv1 calls to klogv2, and since klogv2 is already redirected to log/slog, that should mean both flavors of klog end up in slog in the end

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -634,28 +636,28 @@ func main() {
-	level.Info(logger).Log("host_details", prom_runtime.Uname())
-	level.Info(logger).Log("fd_limits", prom_runtime.FdLimits())
-	level.Info(logger).Log("vm_limits", prom_runtime.VMLimits())
+	logger.Info(version.BuildContext())
+	logger.Info(prom_runtime.Uname())
+	logger.Info(prom_runtime.FdLimits())
+	logger.Info(prom_runtime.VMLimits())
```
I think we should retain the names here, otherwise it is hard to interpret the output.

tjhop replied:
I'll agree with that. They were trimmed initially to fit the output into slog's pre-existing `msg` key. 

Would it be better to log them all as keyvals in a single log call? 

```go
	logger.Info("operational information",
		"build_context", version.BuildContext(),
		"host_details", prom_runtime.Uname(),
		"fd_limits", prom_runtime.FdLimits(),
		"vm_limits", prom_runtime.VMLimits(),
	)
```

tjhop replied:
updated to a single log entry with keyvals, can change further is desired

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -670,37 +672,37 @@ func main() {
-		log.With(logger, "component", "scrape manager"),
-		func(s string) (log.Logger, error) { return logging.NewJSONFileLogger(s) },
+		logger.With("component", "scrape manager"),
+		func(s string) (*logging.JSONFileLogger, error) { return logging.NewJSONFileLogger(s) },
```
Should this return a more abstract interface type?

tjhop replied:
I can be convinced of the usefulness there. 

The previous "raison d'être" for the json file logger, as far as I can tell, was to satisfy the `promql.QueryLogger` interface:
https://github.com/prometheus/prometheus/pull/14906/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaL127-L130

Does that feel more appropriate for our usage here?

bboreham commented on cmd/prometheus/main.go:
```diff
@@ -1430,7 +1433,7 @@ func reloadConfig(filename string, expandExternalLabels, enableExemplarStorage b
 
 	noStepSuqueryInterval.Set(conf.GlobalConfig.EvaluationInterval)
 	l := []interface{}{"msg", "Completed loading of configuration file", "filename", filename, "totalDuration", time.Since(start)}
-	level.Info(logger).Log(append(l, timings...)...)
+	logger.Info(fmt.Sprintf("%v", append(l, timings...)...))
```
This probably needs reworking for slog; it's trying to pass paired name/values.
Perhaps `[]slog.Attr` ?

bboreham reviewed (commented):
Sterling stuff!  Thanks for doing this; some initial comments below.
(I only read about 25% of the PR, so far)

tjhop commented:
@bboreham I believe this is ready for a re-review. I've rebased on current upstream main, incorporated fixes for some of the feedback provided so far, and go tests working locally. 

Things work locally using go workspaces, but as I mention in the updated PR description, this requires functionality in 2 PRs that haven't made their way to an upstream release yet in prometheus/common. Because of this, test failures should be expected as things fail to compile for obvious reasons.

bboreham commented:
Pin to a commit in `go.mod` ?

tjhop commented:
> Pin to a commit in `go.mod` ?

Sure, I can do that. I'd prefer not to do that at merge time, but happy to do that for now at least.

tjhop commented:
All lights are green (after [pinning prometheus/common to commit](https://github.com/prometheus/prometheus/pull/14906/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R55)) :upside_down_face:

SuperQ commented:
I plan to cut a prometheus/common tag after https://github.com/prometheus/common/pull/700.

tjhop commented:
Bumped prometheus/common to v0.60.0, rebased to solve a merge conflict, squashed commits, and updated commit description to match current state :+1:

SuperQ commented:
@roidelapluie Yea, I don't like the full path either. We haven't been able to figure out a way to get the relative path. :disappointed: 

We could change this in the promslog package to only print the filename like we had in go-kit logging.

tjhop commented:
> didn't we go for ts ?

We did in the past and `promslog` will format the timestamp key as `ts` when using the legacy `GoKitStyle` formatting. Since 3.0 presents the opportunity for breaking changes, it was decided to keep it as is and adopt slog as much as possible. Should we revisit that stance?

tjhop commented:
Also, I'm just noticing the lack of a `Z` in the default slog timestamp in that log; doesn't prometheus have an explicit UTC-everywhere policy? If so, I would vote to alter that from the default behavior to use UTC always in our lib

SuperQ commented:
I think we should keep as much of slog format as possible. `time` is fine instead of `ts`.

tjhop commented:
Please excuse the frequent force pushes -- this PR touches a ton of stuff and I'm trying to stay on top of all the merge conflicts, heh

SuperQ reviewed (dismissed):
Awesome work. I think we should merge this asap and iterate on improvements.

tjhop commented:
Agreed, I'd love for this to have some time in a RC before the 3.0 launch

tjhop commented:
resolved another merge conflict and fixed up go mods that got messed up earlier

tjhop commented:
> Mmh, I see slog does not return error, but we sill moved query logger to slog?

Yes, so far we've been operating with the goal of "full slog adoption" as much as possible. The [`QueryLogger` interface](https://github.com/prometheus/prometheus/pull/14906/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaL127-L129) has been updated to match the slog sugar logger methods available, the difference being that `With()` on the query logger does not return a new logger but rather updates the querylogger's internal logger.

I had some [similar thoughts](https://github.com/prometheus/prometheus/pull/14906/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR653-R655). Do you have any thoughts/opinions?

tjhop commented:
I'm game for merge. Like @SuperQ had mentioned a ways above, let's get it in and iterate on improvements :+1:

SuperQ commented:
@roidelapluie As far as I remember looking at the go-kit/log code, the `Log()` interface swallowed all errors anyway and always returned `nil`. I think it was only there for internal use or something like that. This is why we added an exception to golangci-lint configs to ignore `errcheck` issues.

Which reminds me, @tjhop , you can remove this from the `.golanci.yml`

```
      # Never check for logger errors.
      - (github.com/go-kit/log.Logger).Log
```

tjhop commented:
> Which reminds me, @tjhop , you can remove this from the .golanci.yml

Woops, I had it in there at one point, must've lost in one of the recent rebases. Thanks!

tjhop commented:
Did a thorough self-review and see a few other spots to tidy up, will get things better cleaned up
